### ISSUE TITLE
Two-way references and forms don't work together

### ISSUE THREAD

**Issue by fflorent** [gristlabs/grist-core#1280]
### Describe the current behavior

When we try to submit a form bound to a table which uses a two-way reference, we are answered an error. The devtools tells us it is a 403 error.

### Steps to reproduce

You may download this Grist and try to use the form:

Otherwise, follow these steps:

1. Create a new document
2. Create a `Towns` table, with a `name` column and fill some data 
3. Create a `Contacts` table, with a `name` column (text) and a `town` column (reference to the Town table)
4. Configure the `town` column and click on `Add two-way reference` and configure the two-way references with the Towns table
5. Make a new form, publish it, try to submit data using this form

We this error message: 

> There was an error submitting your form. Please try again.

Using the devtools, we see this error in response to the request:

> Blocked by table update access rules

### Describe the expected behavior

We may either:
 - at least display an intelligible error message, telling why the data could not be submitted, and guide how to fix that for the admin;
 - or prevent the form to be 

### Where have you encountered this bug?

- [ ] On [docs.getgrist.com](https://docs.getgrist.com)
- [ ] On a self-hosted instance

### Instance information (when self-hosting only)

* Grist instance:
  * Version:
  * URL (if it's OK for you to share it):
  * Installation mode: docker/kubernetes/...
  * Architecture: single-worker/multi-workers

* Browser name, version and platforms on which you could reproduce the bug:
* Link to browser console log if relevant:
* Link to server log if relevant:

**Comment by dsagal** [comment]
I believe this issue translates to a low-level behavior of access rules, and a question whether it should be relaxed in general or if we need an exception specific to two-way references.

When you change a reference column `C` that's part of a two-way reference, that triggers another change to the reverse column `R` in the target table, where in fact multiple cells may get changed to stay in sync with the change in `C`. In general, forms rely on access rules to restrict an anonymous user's ability to only add a record. In particular, they don't allow anonymous users to change any cells in the target table.

There is a similar situation with formulas: a change to a cell that a user is allowed to edit may lead to calculated update to another cell that the user is not allowed to edit. We have an exception for that -- we call such actions "indirect" actions, and those bypass access rules.

So I see two ways to go about addressing this issue:

1. Consider the updates to a reverse reference column to be "indirect" actions. There is a slightly wider class of such triggered updates (similar to formulas but not the same). To consider them all indirect, you'd wrap the two lines at https://github.com/gristlabs/grist-core/blob/3f49c60e89d9f6bc9f56d610561f8289fdd73d19/sandbox/grist/useractions.py#L526 into `self.indirect_actions()` (and two other similar lines elsewhere).

This would have a more general effect: even without forms, if you permit someone to edit C, but not to edit R (its reverse column), they would be allowed to change R indirectly by changing C. Currently, they would need permission to change both.

2. We could keep the more general behavior strict (if R is a reverse column of C, then to be allowed to change either one, you need to give the user permission to edit both). Then we could address the forms issue by adding that second permission whenever we see that there is a two-way reference, and update it when two-way reference columns are added or removed.

I am leaning toward (1) as the more expected behavior generally. But any relaxing of access rules is a bit hard -- we need to think carefully of unexpected effects and unexpected use-cases that might rely on the current behavior. @paulfitz, what do you think?

**Comment by dsagal** [comment]
FYI, we have a fix in hand, taking approach (1).

**Comment by vviers** [comment]
Has the fix been implemented yet ? A user ran into the same bug today

**Comment by dsagal** [comment]
We weren't satisfied with the fix we had because it turned out that it breaks "undo". It actually solves a more general problem with access rules blocking indirect actions like changes to the other side of a two-way reference. The fix would allow such changes, but an attempt to undo the change would still get blocked. In a way, that seemed worse than disallowing the change in the first place. So we are considering it only a half-fix; the other half would be more work.

For Forms, however, the half-fix would be enough. Perhaps we should land it anyway, and only leave the other half for later -- and perhaps it would be rare for anyone to run into that?

**Comment by vviers** [comment]
> For Forms, however, the half-fix would be enough. Perhaps we should land it anyway, and only leave the other half for later -- and perhaps it would be rare for anyone to run into that?

That sounds good to me, let us know if we can help here !

**Comment by tayflo** [comment]
Is the fix still considered for implementation?

I'm kind fond of two-way references but this bug forbids me to use form if I use them 😕

I guess undoing indirect changes on two-way references is rarer than using a two-way reference column as a form field

**Comment by guillett** [comment]
Hello, I ran into this bug so far I can't deploy a form. (I subscribe to the discussion/thread.)

**Comment by dsagal** [comment]
> > For Forms, however, the half-fix would be enough. Perhaps we should land it anyway, and only leave the other half for later -- and perhaps it would be rare for anyone to run into that?
> 
> That sounds good to me, let us know if we can help here !

Here's what is preventing me from going with this solution. The fix that allows indirect actions to bypass access rules seems correct; what breaks is the undo. However, undo breaks in a bad way: because undo doesn't have full information about what's direct and what's indirect, it applies what it can and skips the rest. Unfortunately, that leads to inconsistent references in case of two-way references. In other words, implementing this fix without addressing the problem with undos will cause inconsistent / corrupt data. That seems entirely unacceptable.

So we need to decide what to do about undos.

- Option 1: allow undos to bypass access rules entirely, under certain conditions, such as: the undo must come from the document history, must be done by the same user who made the original action, and must be of an action done in the last hour (or perhaps a shorter interval)? We have a POC of this working. The downside is that sometimes access rules are intentionally one-way (e.g. a record can be "locked", but cannot be unlocked by the same person). Would it be a problem if undo could revert such changes?

- Option 2: we preserve information about indirect actions and include it into the undo. Then undo can work as today, with access rules still applying to it, but indirect actions can be undone regardless of access rules. This does feel more correct to me, but this would be more work.

Thoughts, anyone?

**Comment by vviers** [comment]
@jonathanperret and I think option 2 looks better as it could happen that users legitimately want some changes to not be undoable, although if it is a lot more work than option 1 I could see the argument for waiting until someone screams about it before doing said extra work

**Comment by dsagal** [comment]
Yes, let's settle on Option 2. It's just more work, so we need to find the time.

I'd like to add another bug report from a user, which I think has the same root cause. If a column is configured so a specific user read-only access to it, but has structure permission (so can edit formulas), then they can edit the formula in the column, but get blocked trying to undo that edit. I think the root cause here is also that the "undo" operation does not know which actions in the undo bundle should be considered direct and which indirect.

**Comment by QurQuma** [comment]
this is not happening if form is not published yet
are there info about this fix release?
this issue definitely prevent me from using cross references and grist in general.
thanks

**Comment by Erlie14** [comment]
Hello, I have the same problem when I use lookupAddDerived. I can't submit the form. I would like to kwnow if you have a solution for this problem.  Thank you.
